src/vec/FixedPairList.cpp

```diff
@@ -0,0 +1,336 @@
+  }
+
+  python::list FixedPairList::getBonds()
+  {
+    python::tuple bond;
+    python::list bonds;
+    for (auto it = globalPairs.cbegin(); it != globalPairs.cend(); it++) {
+      bond = python::make_tuple(it->first, it->second);
+      bonds.append(bond);
+    }
+
+    return bonds;
+  }
+
+  std::vector<size_t> FixedPairList::getPairList()
+  {
+    std::vector<size_t> ret;
+    for (auto it = globalPairs.cbegin(); it != globalPairs.cend(); it++) {
+      ret.push_back(it->first);
+      ret.push_back(it->second);
+    }
+    return ret;
+  }
+
+  python::list FixedPairList::getAllBonds()
+  {
+    std::vector<size_t> local_bonds = getPairList();
+    std::vector<std::vector<size_t> > global_bonds;
+    python::list bonds;
+
+    for (auto it = globalPairs.cbegin(); it != globalPairs.cend(); it++) {
+      local_bonds.push_back(it->first);
+      local_bonds.push_back(it->second);
+    }
+    System& system = vectorization->getSystemRef();
+    if (system.comm->rank() == 0) {
+      mpi::gather(*system.comm, local_bonds, global_bonds, 0);
+      for (auto it = global_bonds.begin(); it != global_bonds.end(); ++it) {
+        for (auto iit = it->begin(); iit != it->end(); iit += 2) {
+          size_t pid1 = *(iit);
+          size_t pid2 = *(iit + 1);
+          bonds.append(python::make_tuple(pid1, pid2));
+        }
+      }
+    } else {
+      mpi::gather(*system.comm, local_bonds, global_bonds, 0);
+    }
+    return bonds;
+  }
+
+  void FixedPairList::beforeSendParticles(ParticleList& pl, OutBuffer& buf)
+  {
+    std::vector< size_t > toSend;
+    // loop over the particle list
+    for (ParticleList::Iterator pit(pl); pit.isValid(); ++pit) {
+      size_t pid = pit->id();
+
+      LOG4ESPP_DEBUG(theLogger, "send particle with pid " << pid << ", find pairs");
+
+      // find all pairs that involve this particle
+
+      int n = globalPairs.count(pid);
+
+      if (n > 0) {
+        const auto equalRange = globalPairs.equal_range(pid);
+
+        // first write the pid of the first particle
+        // then the number of partners
+        // and then the pids of the partners
+        toSend.reserve(toSend.size()+n+1);
+        toSend.push_back(pid);
+        toSend.push_back(n);
+        for (auto it = equalRange.first; it != equalRange.second; ++it) {
+          toSend.push_back(it->second);
+          LOG4ESPP_DEBUG(theLogger, "send global bond: pid "
+                       << pid << " and partner " << it->second);
+        }
+
+        // delete all of these pairs from the global list
+        globalPairs.erase(equalRange.first, equalRange.second);
+        // std::cout << "erasing particle " << pid << " from here" << std::endl;
+      }
+    }
+    // send the list
+    buf.write(toSend);
+    LOG4ESPP_INFO(theLogger, "prepared fixed pair list before send particles");
+  }
+
+  void FixedPairList::afterRecvParticles(ParticleList &pl, InBuffer& buf)
+  {
+    std::vector< size_t > received;
+    int n;
+    size_t pid1, pid2;
```

Comment:
Do not use uninitialized variables! Possible source of undefined behavior. Declare them where you need them.